File: problems/quest/data_structures_and_algorithms/arrays/Set_Mismatch.py
from typing import List


class SetMismatch:
    # Очень долго допирал, что и как сделать (19ms)
    def findErrorNums1(self, nums: List[int]) -> List[int]:
        sorted_nums = sorted(nums)
        ans = []

        if len(nums) == 2:
            if nums[0] == 1:
                ans = [1, 2]
            else:
                ans = [2, 1]
            return ans

        last_num = 0
        for num in sorted_nums:
            if num == last_num:
                ans.append(num)
            last_num = num

        set_nums = sorted_nums.copy()
        set_nums.remove(ans[0])

        for i in range(len(set_nums)):
            if set_nums[i] != i + 1:
                ans.append(i + 1)
                return ans

        ans.append(len(nums))
        return ans

    # Лучшее и понятное решение через количество вхождений (11ms)
    def findErrorNums2(self, nums: List[int]) -> List[int]:
        n = len(nums)
        # Нулевой элемент не используется и нужен для обращения по индексам с единицы
        frequency = [0] * (n + 1)

        for num in nums:
            frequency[num] += 1

        duplicate = 0
        missing = 0
        for i in range(1, n + 1):
            if frequency[i] == 2:
                duplicate = i
            if frequency[i] == 0:
                missing = i

        return [duplicate, missing]

File: problems/quest/data_structures_and_algorithms/arrays/SetMismatch.java
import java.util.Arrays;

public class SetMismatch {
    // 14ms
    public int[] findErrorNums1(int[] nums) {
        Arrays.sort(nums);
        // Исключительные случаи [1,1] и [2,2]
        if (nums.length == 2) {
            if (nums[0] == 1) {
                return new int[] {1,2};
            } else {
                return new int[] {2,1};
            }
        }

        int[] ans = new int[2];
        // Ищем повторяющийся элемент и его индекс
        int last_num = 0;
        int last_num_idx = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == last_num) {
                ans[0] = nums[i];
                last_num_idx = i;
            }
            last_num = nums[i];
        }

        // Создаем "сет" уникальных элементов
        int[] set_nums = new int[nums.length - 1];
        for (int i = 0; i < last_num_idx; i++)
            set_nums[i] = nums[i];
        for (int i = last_num_idx; i < set_nums.length; i++)
            set_nums[i] = nums[i + 1];

        // Проходимся по сету в поисках отсутствующего элемента
        for (int j = 0; j < set_nums.length; j++) {
            if (set_nums[j] != j + 1) {
                ans[1] = j + 1;
                
                return ans;
            }
        }
        // Если отсутствующий элемент не был найден - это последний элемент массива
        ans[1] = nums.length;

        return ans;
    }
    // 2ms
    public int[] findErrorNums2(int[] nums) {
        int n = nums.length;
        int[] frequency = new int[n + 1];

        for (int num : nums)
            frequency[num]++;

        int duplicate = 0, missing = 0;
        for (int i = 1; i <= n; i++) {
            if (frequency[i] == 2)
                duplicate = i;
            if (frequency[i] == 0)
                missing = i;
        }

        return new int[] {duplicate, missing};
    }
}